src/gorilla.c

```diff
@@ -0,0 +1,451 @@
+static binary_t readBits(binary_t *bins, globalbit_t *bit, u_int8_t dataLen) {
+    binary_t *bin_it = Bins_bitbin(bins, *bit);
+    localbit_t lbit = localbit(*bit);
+    localbit_t available = BINW - lbit;
+
+    binary_t bin = 0;
+    if (available >= dataLen) {
+        bin = LSB(*bin_it >> lbit, dataLen);
+    } else {
+        u_int8_t left = dataLen - available;
+        bin = LSB(*bin_it >> lbit, available);
+        bin |= LSB(*++bin_it, left) << available;
+    }
+    *bit += dataLen;
+    return bin;
+}
+
+static bool isSpaceAvailable(CompressedChunk *chunk, u_int8_t size) {
+    u_int64_t available = (chunk->size * 8) - chunk->idx;
+    return size <= available;
+}
+
+/***************************** APPEND ********************************/
+static ChunkResult appendTS(CompressedChunk *chunk, timestamp_t timestamp) {
+    assert(timestamp >= chunk->prevTimestamp);
+    timestamp_t curDelta = timestamp - chunk->prevTimestamp;
+    
+    union64bits doubleDelta;
+    doubleDelta.i = curDelta - chunk->prevTimestampDelta;
+/* 
+ * Before any insertion the code `CHECKSPACE` ensures there is enough space to
+ * encode timestamp and one additional bit which the minimum to encode the value.
+ * This is why we have `+ 1` in `CHECKSPACE`.
+ * 
+ * If doubleDelta == 0, 1 bit of value 0 is inserted.
+ * 
+ * Else, `Bin_InRange` checks for the minimal number of bits required to represent
+ * `doubleDelta`, the delta of deltas between current and previous timestamps.
+ * Then two values are being inserted. 
+   * The first value is, encoding for the lowest number of bits for which
+     `Bin_InRange` returns `true`.
+   * The second value is a compressed representation of the value with the `length`
+     encoded by the first value. Compression is done using `int2bin`.
+ */
+    u_int64_t *bins = chunk->data;
+    u_int64_t *bit = &chunk->idx;
+    if (doubleDelta.i == 0) {
+        CHECKSPACE(chunk, 1 + 1); // CHECKSPACE adds 1 as minimum for double space
+        appendBits(bins, bit, 0x00, 1);
+    } else if (Bin_InRange(doubleDelta.i, 7)) {
+        CHECKSPACE(chunk, 2 + 7 + 1);
+        appendBits(bins, bit, 0x01, 2);
+        appendBits(bins, bit, int2bin(doubleDelta.i, 7), 7);
+    } else if (Bin_InRange(doubleDelta.i, 10)) {
+        CHECKSPACE(chunk, 3 + 10 + 1);
+        appendBits(bins, bit, 0x03, 3);
+        appendBits(bins, bit, int2bin(doubleDelta.i, 10), 10);
+    } else if (Bin_InRange(doubleDelta.i, 13)) {
+        CHECKSPACE(chunk, 4 + 13 + 1);
+        appendBits(bins, bit, 0x07, 4);
+        appendBits(bins, bit, int2bin(doubleDelta.i, 13), 13);
+    } else if (Bin_InRange(doubleDelta.i, 16)) {
+        CHECKSPACE(chunk, 5 + 16 + 1);
+        appendBits(bins, bit, 0x0f, 5);
+        appendBits(bins, bit, int2bin(doubleDelta.i, 16), 16);
+    } else {
+        CHECKSPACE(chunk, 5 + 64 + 1);
+        appendBits(bins, bit, 0x1f, 5);
+        appendBits(bins, bit, doubleDelta.u, 64);
+    }
+    chunk->prevTimestampDelta = curDelta;
```

Comment:
I think this should move to `CChunk_Append`, if appenV fails you don't want to change `prevTimestampDelta` and `prevTimestamp` just yet, only after you fully wrote the entire sample.

Reply:
If appending the value will fail, these bits will not be read since the count of samples in the chunk will not increase by 1.